modules/features/clip/src/main/kotlin/au/com/shiftyjelly/pocketcasts/clip/ClipSelector.kt

```diff
@@ -57,24 +89,289 @@ internal fun ClipSelector(
                 .size(72.dp)
                 .clip(RoundedCornerShape(topStart = 8.dp, bottomStart = 8.dp))
                 .clickable(
-                    interactionSource = remember { MutableInteractionSource() },
+                    interactionSource = remember(::MutableInteractionSource),
                     indication = rememberRipple(color = clipColors.baseColor),
                     onClickLabel = stringResource(if (isPlaying) LR.string.pause else LR.string.play),
                     role = Role.Button,
                     onClick = if (isPlaying) onPauseClick else onPlayClick,
                 )
                 .padding(16.dp),
         )
+        Spacer(
+            modifier = Modifier.width(16.dp),
+        )
+        ClipSelector(
+            episodeDuration = episodeDuration,
+            clipRange = clipRange,
+            clipColors = clipColors,
+            onClipStartUpdate = onClipStartUpdate,
+            onClipEndUpdate = onClipEndUpdate,
+            state = state,
+        )
+    }
+}
+
+@Composable
+private fun ClipSelector(
+    episodeDuration: Duration,
+    clipRange: Clip.Range,
+    clipColors: ClipColors,
+    state: ClipSelectorState,
+    onClipStartUpdate: (Duration) -> Unit,
+    onClipEndUpdate: (Duration) -> Unit,
+) {
+    BoxWithConstraints(
+        modifier = Modifier.fillMaxSize(),
+    ) {
+        ClipTimeline(
+            episodeDuration = episodeDuration,
+            clipRange = clipRange,
+            clipColors = clipColors,
+            state = state,
+            onClipStartUpdate = onClipStartUpdate,
+            onClipEndUpdate = onClipEndUpdate,
+        )
+        ClipBox(
+            episodeDuration = episodeDuration,
+            clipRange = clipRange,
+            clipColors = clipColors,
+            state = state,
+            onClipStartUpdate = onClipStartUpdate,
+            onClipEndUpdate = onClipEndUpdate,
+        )
+    }
+}
+
+@Composable
+private fun BoxWithConstraintsScope.ClipTimeline(
+    episodeDuration: Duration,
+    clipRange: Clip.Range,
+    clipColors: ClipColors,
+    state: ClipSelectorState,
+    onClipStartUpdate: (Duration) -> Unit,
+    onClipEndUpdate: (Duration) -> Unit,
+) {
+    val localDensity = LocalDensity.current
+    val largeTickHeight = maxHeight / 3
+    val mediumTickHeight = maxHeight / 6
+    val smallTickHeight = maxHeight / 12
+
+    val tickWidth = 1.5.dp
+    val spaceWidth = 4.dp * state.scale
+    val itemWidth = tickWidth + spaceWidth
+    state.itemWidth = with(localDensity) { itemWidth.toPx() }
+    LaunchedEffect(state.scale) { state.scaleBoxOffsets(clipRange) }
```

Comment:
I'm not super happy about this part but I think it is a chicken and egg problem. I need `itemWidth` to correctly translate duration into pixels but an actual value is not available before the view is laid out. So I just update it whenever possible and let Compose handle the rest by propagating the state.